system-parachains/coretime/coretime-polkadot/src/coretime.rs

```diff
@@ -0,0 +1,307 @@
+}
+
+/// Burn revenue from coretime sales. See
+/// [RFC-010](https://polkadot-fellows.github.io/RFCs/approved/0010-burn-coretime-revenue.html).
+pub struct BurnCoretimeRevenue;
+impl OnUnbalanced<Credit<AccountId, Balances>> for BurnCoretimeRevenue {
+	fn on_nonzero_unbalanced(amount: Credit<AccountId, Balances>) {
+		let acc = CoretimeBurnAccount::get();
+		if !System::<Runtime>::account_exists(&acc) {
+			// The account doesn't require ED to survive.
+			System::<Runtime>::inc_providers(&acc);
+		}
+		Balances::resolve(&acc, amount).defensive_ok();
+	}
+}
+
+type AssetTransactor = <xcm_config::XcmConfig as xcm_executor::Config>::AssetTransactor;
+
+fn burn_at_relay(stash: &AccountId, value: Balance) -> Result<(), XcmError> {
+	let dest = Location::parent();
+	let stash_location =
+		Junction::AccountId32 { network: None, id: stash.clone().into() }.into_location();
+	let asset = Asset { id: AssetId(Location::parent()), fun: Fungible(value) };
+	let dummy_xcm_context = XcmContext { origin: None, message_id: [0; 32], topic: None };
+
+	let withdrawn = AssetTransactor::withdraw_asset(&asset, &stash_location, None)?;
+
+	// TODO https://github.com/polkadot-fellows/runtimes/issues/404
+	AssetTransactor::can_check_out(&dest, &asset, &dummy_xcm_context)?;
+
+	let parent_assets = Into::<Assets>::into(withdrawn)
+		.reanchored(&dest, &Here)
+		.defensive_map_err(|_| XcmError::ReanchorFailed)?;
+
+	PolkadotXcm::send_xcm(
+		Here,
+		Location::parent(),
+		Xcm(vec![
+			Instruction::UnpaidExecution {
+				weight_limit: WeightLimit::Unlimited,
+				check_origin: None,
+			},
+			ReceiveTeleportedAsset(parent_assets.clone()),
+			BurnAsset(parent_assets),
+		]),
+	)?;
+
+	AssetTransactor::check_out(&dest, &asset, &dummy_xcm_context);
+
+	Ok(())
+}
+
+parameter_types! {
+	/// The revenue from on-demand coretime sales. This is distributed amonst those who contributed
+	/// regions to the pool.
+	pub storage CoretimeRevenue: Option<(BlockNumber, Balance)> = None;
+}
+
+/// Type that implements the [`CoretimeInterface`] for the allocation of Coretime. Meant to operate
+/// from the parachain context. That is, the parachain provides a market (broker) for the sale of
+/// coretime, but assumes a `CoretimeProvider` (i.e. a Relay Chain) to actually provide cores.
+pub struct CoretimeAllocator;
+impl CoretimeInterface for CoretimeAllocator {
+	type AccountId = AccountId;
+	type Balance = Balance;
+	type RelayChainBlockNumberProvider = RelaychainDataProvider<Runtime>;
+
+	fn request_core_count(count: CoreIndex) {
+		use crate::coretime::CoretimeProviderCalls::RequestCoreCount;
+		let request_core_count_call = RelayRuntimePallets::Coretime(RequestCoreCount(count));
+
+		// Weight for `request_core_count` from Polkadot runtime benchmarks:
+		// `ref_time` = 7889000 + (3 * 25000000) + (1 * 100000000) = 182889000
+		// `proof_size` = 1636
+		// Add 5% to each component and round to 2 significant figures.
+		// TODO check when benchmarks are rerun
```

Comment:
Who is checking these TODOs?

Reply:
I will check them, but if the hardcoded weights are not sufficient for the new weights they will make the emulated tests fail anyway.